### PR TITLE
Core C# SDK Rewrite - The Big One

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
Soooo...... here we are. Finally at last. Many many many changes. Everything has been merged with latest master
#### Tests
Ran everything locally - it all worked fine. (Minus a final bug with the Core c# SDK which should be ready by EoD - will push the new package before merging)
#### Documentation
Documentation to follow (UTY-963)
#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.
@zeroZshadow

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,13 +1,25 @@
 /build/assembly/
 /build/descriptor/
 /build/build_logs/
+<<<<<<< HEAD
```
Comment (jessicafalk) [inline .gitignore]: baaaad

Comment (jamiebrynes7) [reply .gitignore]: WAT

Comment (jamiebrynes7) [reply .gitignore]: No one ever saw this 👀

File: README.md [hunk 2]
```diff
@@ -1,3 +1,14 @@
+# The SpatialOS CSharp Core SDK readme (EXPERIMENTAL)
```
Comment (jessicafalk) [inline README.md]: do we want to keep that huge warning?

Comment (jamiebrynes7) [reply README.md]: I'm so confused - I checked out master HEAD for the docs :/

File: ci/download-core-sdk.sh [hunk 3]
```diff
@@ -56,4 +60,8 @@ rm tools/schema_compiler/win/protoc.exe
 rm tools/schema_compiler/macos/protoc
 rm -rf tools/schema_compiler/win/proto
 rm -rf tools/schema_compiler/macos/proto
-rm -rf "${CORE_SDK_DIR}"
+
+# Remove header files in C API
+rm -rf "${NATIVE_DEPENDENCIES_PATH}/Windows/x86_64/include"
+rm -rf "${NATIVE_DEPENDENCIES_PATH}/Linux/x86_64/include"
+rm -rf "${NATIVE_DEPENDENCIES_PATH}/OSX/include"
```
Comment (jessicafalk) [inline ci/download-core-sdk.sh]: new line

File: ci/test.sh [hunk 4]
```diff
@@ -48,6 +48,7 @@ pushd "workers/unity"
     -testPlatform editmode \
     -logfile "${PROJECT_DIR}/logs/unity-editmode-test-run.log" \
     -testResults "${EDITMODE_TEST_RESULTS_FILE}"
+popd
```
Comment (firtina-improbable) [inline ci/test.sh]: looks like a semantic merge conflict, is this supposed to be added here?

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/TransformSendSystem.cs [hunk 5]
```diff
@@ -43,12 +45,11 @@ protected override void OnUpdate()
...
-                update.SetLocation(global::Generated.Improbable.Transform.Location.ToSpatial(component.Location));
-                update.SetRotation(global::Generated.Improbable.Transform.Quaternion.ToSpatial(component.Rotation));
-                update.SetTick(component.Tick);
-                Generated.Improbable.Transform.Transform.Translation.SendComponentUpdate(Worker.Connection, entityId,
-                    update);
+
+                var update = new SchemaComponentUpdate(component.ComponentId);
+                Generated.Improbable.Transform.SpatialOSTransform.Serialization.Serialize(component,
+                    update.GetFields());
+                worker.Connection.SendComponentUpdate(entityId, new ComponentUpdate(update));
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/TransformSendSystem.cs]: Was looking at random files for curiosity, should this be using the raw connection bits?

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/TransformSendSystem.cs]: Yes, currently. There's a ticket open to have helper methods for these.

**Comment by jessicafalk** [comment]
Still getting `InvalidOperationException: Collection was modified` when quitting play mode. Seems like something doesn't get disposed correctly when destroying the `SpatialOSReceiveSystem`

`InvalidOperationException: Collection was modified; enumeration operation may not execute.
System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () (at <f826c2584fc94ec19a48a6576640bdc5>:0)
System.Linq.Enumerable+WhereSelectEnumerableIterator`2[TSource,TResult].MoveNext () (at <839a3cb835c04d14aeb58d83bb7bc4bd>:0)
Generated.Playground.CubeColor+ReferenceTypeProviders+ChangeColorProvider.CleanDataInWorld (Unity.Entities.World world) (at <f4e7a5811d424408b8b9a531f4692871>:0)
Generated.Playground.CubeColor+DispatcherHandler.Dispose () (at <f4e7a5811d424408b8b9a531f4692871>:0)
Improbable.Gdk.Core.SpatialOSReceiveSystem.OnDestroyManager () (at <0ce2f167d2e0497ab48bb7091dc402c9>:0)
Unity.Entities.ScriptBehaviourManager.DestroyInstance () (at <b1af3600ef274b149df519a0a0be698d>:0)
Unity.Entities.World.Dispose () (at <b1af3600ef274b149df519a0a0be698d>:0)
UnityEngine.Debug:LogException(Exception)
Unity.Debug:LogException(Exception)
Unity.Entities.World:Dispose()
Improbable.Gdk.Core.Worker:Dispose()
Playground.Bootstrap:DomainUnloadShutdown()`

**Pull request merged**